Simplenote/SPNoteHistoryController.swift

```diff
@@ -0,0 +1,140 @@
+import Foundation
+
+// MARK: - SPNoteHistoryController: Business logic for history view controller
+//
+final class SPNoteHistoryController {
+
+    // MARK: - Presentable: history item representation for a view
+    //
+    struct Presentable {
+        let date: String
+        let isRestorable: Bool
+    }
+
+    // MARK: - State of the screen
+    //
+    enum State {
+        /// Loading data
+        ///
+        case loading
+
+        /// Ready to show results
+        ///
+        case results([Presentable])
+
+        /// Error
+        ///
+        case error(String)
+    }
+
+    // MARK: - Event which is sent to a delegate
+    //
+    enum Event {
+        /// Request to be dismissed
+        ///
+        case dismiss
+
+        /// Preview version content
+        ///
+        case preview(String)
+
+        /// Restore content to selected version
+        ///
+        case restore
+    }
+
+    /// Observer sends changes of the state (to history view controller)
+    /// When assigned, it sends current state
+    ///
+    var observer: ((State) -> Void)? {
+        didSet {
+            observer?(state)
+        }
+    }
+
+    /// Delegate
+    ///
+    var delegate: ((Event) -> Void)?
```

Comment:
May I suggest replacing the closure / enum pattern, with an old school `protocol`?

Reason is just avoiding the switch in the handling closure. Since handlers for each one of the Event cases are mandatory, the code gets much simpler, and improves readability as well.

Reply:
I think both ways have their benefits. 
Some thoughts:
* With closure it is possible to have a private implementation of delegate methods. When a class implements a protocol all methods are exposed to the "outside", which clutters the code completion a bit.
* Traditionally names of the functions in a protocol are quite verbose and long, including contextual information, like  name and instance of the object. For example `func noteHistoryControllerDidCancel(_ controller: SPNoteHistoryController)`. With closure contextual information is often inferred and names can be more concise.

I changed the current implementation from the closure to a protocol to follow the project 👍